contracts/liquidity_hub/bonding-manager/src/contract.rs

```diff
@@ -0,0 +1,295 @@
+            )
+        }
+        ExecuteMsg::Unbond { asset } => {
+            cw_utils::nonpayable(&info)?;
+            commands::unbond(deps, env.block.time, info, env, asset)
+        }
+        ExecuteMsg::Withdraw { denom } => {
+            cw_utils::nonpayable(&info)?;
+            commands::withdraw(deps, env.block.time, info.sender, denom)
+        }
+        ExecuteMsg::UpdateConfig {
+            owner,
+            pool_manager_addr,
+            unbonding_period,
+            growth_rate,
+        } => commands::update_config(
+            deps,
+            info,
+            owner,
+            pool_manager_addr,
+            unbonding_period,
+            growth_rate,
+        ),
+        ExecuteMsg::FillRewards { .. } => commands::fill_rewards(deps, env, info),
+        ExecuteMsg::FillRewardsCoin => commands::fill_rewards(deps, env, info),
+        ExecuteMsg::Claim { .. } => commands::claim(deps, env, info),
+        ExecuteMsg::EpochChangedHook { current_epoch } => {
+            // Epoch has been updated, update rewards bucket
+            // and forward the expiring epoch
+            // Store epoch manager and verify the sender is him
+            println!("New epoch created: {:?}", current_epoch);
+
+            let new_epoch_id = current_epoch.id;
+            let next_epoch_id = new_epoch_id.checked_add(1u64).unwrap();
+            // Creates a new bucket for the rewards flowing from this time on, i.e. to be distributed in the next epoch. Also, forwards the expiring epoch (only 21 epochs are live at a given moment)
+            // Add a new rewards bucket for the new epoch
+            EPOCHS.save(
+                deps.storage,
+                &next_epoch_id.to_be_bytes(),
+                &Epoch {
+                    id: next_epoch_id.into(),
+                    start_time: current_epoch.start_time.plus_days(1),
+                    ..Epoch::default()
+                },
+            )?;
+            println!("New epoch created: {}", next_epoch_id);
+            // Return early if the epoch is the first one
+            if new_epoch_id == 1 {
+                // Creates a new bucket for the rewards flowing from this time on, i.e. to be distributed in the next epoch. Also, forwards the expiring epoch (only 21 epochs are live at a given moment)
+                // Add a new rewards bucket for the new epoch
+                EPOCHS.save(
+                    deps.storage,
+                    &new_epoch_id.to_be_bytes(),
+                    &Epoch {
+                        id: next_epoch_id.into(),
+                        start_time: current_epoch.start_time,
+                        ..Epoch::default()
+                    },
+                )?;
+                return Ok(Response::default()
+                    .add_attributes(vec![("action", "epoch_changed_hook".to_string())]));
+            }
+
+            let expiring_epoch_id = new_epoch_id.checked_sub(1u64).unwrap();
```

Comment:
This should be subtracting 21, not 1. And that's only to be done if the epoch_id is higher than 21, otherwise there's no point doing it.